# Exercício *Gasolina Do Brasil.*

## Neste projeto, trabalharemos com dois arquivos csvs separados, contendo informações sobre o preço da gasolina no Brasil. 

### 1. Carregue os conjuntos de dados "gasolina_2000+.csv" e "gasolina_2010+.csv" em dois DataFrames diferentes e combine-os em um único DataFrame.

In [112]:
import pandas as pd

In [113]:
df_1 = pd.read_csv('C:/Users/55859/Desktop/Estudos/Data Science/02 Analisando dados com Pandas/DataSetes/gasolina_2000+.csv')

In [114]:
df_2 = pd.read_csv('C:/Users/55859/Desktop/Estudos/Data Science/02 Analisando dados com Pandas/DataSetes/gasolina_2010+.csv')

C:\Users\55859\AppData\Local\Temp\ipykernel_5940\2508353403.py:1: DtypeWarning: Columns (12,14,15,16,17,18) have mixed types. Specify dtype option on import or set low_memory=False.
  df_2 = pd.read_csv('C:/Users/55859/Desktop/Estudos/Data Science/02 Analisando dados com Pandas/DataSetes/gasolina_2010+.csv')


### 2.Investigue as colunas e entenda o conjunto de dados usando o head() e info()

In [115]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42638 entries, 0 to 42637
Data columns (total 19 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     42638 non-null  int64  
 1   DATA INICIAL                   42638 non-null  object 
 2   DATA FINAL                     42638 non-null  object 
 3   REGIÃO                         42638 non-null  object 
 4   ESTADO                         42638 non-null  object 
 5   PRODUTO                        42638 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   42638 non-null  int64  
 7   UNIDADE DE MEDIDA              42638 non-null  object 
 8   PREÇO MÉDIO REVENDA            42638 non-null  float64
 9   DESVIO PADRÃO REVENDA          42638 non-null  float64
 10  PREÇO MÍNIMO REVENDA           42638 non-null  float64
 11  PREÇO MÁXIMO REVENDA           42638 non-null  float64
 12  MARGEM MÉDIA REVENDA           42638 non-null 

In [116]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78185 entries, 0 to 78184
Data columns (total 19 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     78185 non-null  int64  
 1   DATA INICIAL                   78185 non-null  object 
 2   DATA FINAL                     78185 non-null  object 
 3   REGIÃO                         78185 non-null  object 
 4   ESTADO                         78185 non-null  object 
 5   PRODUTO                        78185 non-null  object 
 6   NÚMERO DE POSTOS PESQUISADOS   78185 non-null  int64  
 7   UNIDADE DE MEDIDA              78185 non-null  object 
 8   PREÇO MÉDIO REVENDA            78185 non-null  float64
 9   DESVIO PADRÃO REVENDA          78185 non-null  float64
 10  PREÇO MÍNIMO REVENDA           78185 non-null  float64
 11  PREÇO MÁXIMO REVENDA           78185 non-null  float64
 12  MARGEM MÉDIA REVENDA           78185 non-null 

#### Agora vamos juntar os dois arquivos em um só.

In [117]:
df_1.head(3)

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO            ESTADO  \
0           0   2004-05-09  2004-05-15  CENTRO OESTE  DISTRITO FEDERAL   
1           1   2004-05-09  2004-05-15  CENTRO OESTE             GOIAS   
2           2   2004-05-09  2004-05-15  CENTRO OESTE       MATO GROSSO   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                  1.19   
1                1.162                  0.114                  0.89   
2                1.389                  0.097                  1.18   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  
0                         0.133  
1                         0.115  
2                         0.098

In [118]:
df_2.head(3)

Unnamed: 0 DATA INICIAL  DATA FINAL        REGIÃO            ESTADO  \
0        9256   2011-01-02  2011-01-08  CENTRO OESTE  DISTRITO FEDERAL   
1        9257   2011-01-02  2011-01-08  CENTRO OESTE             GOIAS   
2        9258   2011-01-02  2011-01-08  CENTRO OESTE       MATO GROSSO   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                            80              R$/l   
1  ETANOL HIDRATADO                           255              R$/l   
2  ETANOL HIDRATADO                           141              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                2.030                  0.003                  2.02   
1                1.717                  0.102                  1.49   
2                1.848                  0.148                  1.59   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                  2.05                0.411                     0.001   
1                  1.99                0.225                     0.059   
2                  2.49                0.326                     0.080   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    1.619                      0.013   
1                    1.492                      0.069   
2                    1.522                      0.105   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    1.6111                    1.6411   
1                    1.3221                    1.6201   
2                    1.2833                    1.7486   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  
0                         0.008  
1                         0.046  
2                         0.069

### 3. Selecione a terceira entrada da coluna DATA INICIAL e verifique seu tipo.

##### Perceba que é interessante notar que os dois métodos retornam o valor da coluna correspondente ao que está dentro de loc ou iloc.

In [119]:
df_1['NÚMERO DE POSTOS PESQUISADOS'].loc[2]

192

In [120]:
type(df_1['DATA INICIAL'])

pandas.core.series.Series

In [121]:
type(df_1['DATA INICIAL'].iloc[0])

str

In [122]:
df_2['NÚMERO DE POSTOS PESQUISADOS'].iloc[2]

141

In [123]:
type(df_2['DATA INICIAL'])

pandas.core.series.Series

In [124]:
type(df_2['DATA INICIAL'].iloc[0])

str

### 4. Você deve ter percebido que as colunas DATA INICIAL e DATA FINAL estão formatadas como string. Utilizando o método pd.to_datetime(), converta ambas colunas para Timestamp / Datetime.

In [125]:
df_1['DATA INICIAL'] = pd.to_datetime(df_1['DATA INICIAL'])

In [126]:
df_1['DATA FINAL'] = pd.to_datetime(df_1['DATA FINAL'])

In [127]:
type(df_1['DATA FINAL'].iloc[0])

pandas._libs.tslibs.timestamps.Timestamp

In [128]:
df_2['DATA INICIAL'] = pd.to_datetime(df_2['DATA INICIAL'])

In [129]:
df_2['DATA FINAL'] = pd.to_datetime(df_2['DATA FINAL'])

In [130]:
type(df_2['DATA FINAL'].iloc[0])

pandas._libs.tslibs.timestamps.Timestamp

In [131]:
df = pd.concat([df_1, df_2])
df.head(5)

Unnamed: 0 DATA INICIAL DATA FINAL        REGIÃO              ESTADO  \
0           0   2004-05-09 2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1           1   2004-05-09 2004-05-15  CENTRO OESTE               GOIAS   
2           2   2004-05-09 2004-05-15  CENTRO OESTE         MATO GROSSO   
3           3   2004-05-09 2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4           4   2004-05-09 2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                  1.19   
1                1.162                  0.114                  0.89   
2                1.389                  0.097                  1.18   
3                1.262                  0.070                  1.09   
4                1.181                  0.078                  1.05   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   
3                 1.509                0.432                     0.055   
4                 1.400                 0.24                     0.066   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   
3                     0.83                      0.119   
4                    0.941                      0.077   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   
3                    0.5991                   1.22242   
4                    0.7441                    1.0317   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  
0                         0.133  
1                         0.115  
2                         0.098  
3                         0.143  
4                         0.082

### 5. Crie uma nova coluna para representar o ano e o mês(aaaa-mm), utilizando a coluna DATA FINAL como referência. 

In [132]:
df['ANO - MÊS'] = df['DATA FINAL'].apply(lambda ano : "{}".format(ano.year)) + df['DATA FINAL'].apply(lambda mes : " - {:02d}".format(mes.month))

In [133]:
df.head(3)

Unnamed: 0 DATA INICIAL DATA FINAL        REGIÃO            ESTADO  \
0           0   2004-05-09 2004-05-15  CENTRO OESTE  DISTRITO FEDERAL   
1           1   2004-05-09 2004-05-15  CENTRO OESTE             GOIAS   
2           2   2004-05-09 2004-05-15  CENTRO OESTE       MATO GROSSO   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                  1.19   
1                1.162                  0.114                  0.89   
2                1.389                  0.097                  1.18   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  ANO - MÊS  
0                         0.133  2004 - 05  
1                         0.115  2004 - 05  
2                         0.098  2004 - 05

### 6. Utilizando o value_counts(), liste todos os tipos de produtos contidos na base de dados.'m

In [134]:
df['PRODUTO'].value_counts()

GASOLINA COMUM        23570
GLP                   23561
ETANOL HIDRATADO      23440
ÓLEO DIESEL           21194
GNV                   14469
ÓLEO DIESEL S10        9113
OLEO DIESEL S10        2376
OLEO DIESEL            2351
GASOLINA ADITIVADA      749
Name: PRODUTO, dtype: int64

In [135]:
df['REGIÃO'].value_counts()

NORDESTE        41671
NORTE           27989
SUDESTE         19270
CENTRO OESTE    17449
SUL             14444
Name: REGIÃO, dtype: int64

In [136]:
df['ESTADO'].value_counts()

PARAIBA                4819
RIO DE JANEIRO         4819
SAO PAULO              4819
RIO GRANDE DO SUL      4819
BAHIA                  4819
SANTA CATARINA         4818
RIO GRANDE DO NORTE    4818
MINAS GERAIS           4817
ESPIRITO SANTO         4815
PERNAMBUCO             4813
CEARA                  4813
PARANA                 4807
ALAGOAS                4806
SERGIPE                4795
MATO GROSSO DO SUL     4793
MATO GROSSO            4566
AMAZONAS               4414
GOIAS                  4134
PIAUI                  4017
MARANHAO               3971
DISTRITO FEDERAL       3956
PARA                   3955
TOCANTINS              3953
RONDONIA               3950
ACRE                   3948
RORAIMA                3938
AMAPA                  3831
Name: ESTADO, dtype: int64

### 7. Filtre o DataFrame para obter apenas dados da 'GASOLINA COMUM'. Grave em uma nova variável.
#### Detalhe importante é que eu poderia ter pego qualquer um deses:
    GASOLINA COMUM        
    GLP                   
    ETANOL HIDRATADO      
    ÓLEO DIESEL           
    GNV                   
    ÓLEO DIESEL S10      
    OLEO DIESEL S10   
    OLEO DIESEL         
    GASOLINA ADITIVADA    

In [137]:
df_amostra = df[df['PRODUTO'] == 'GASOLINA COMUM']
df_amostra

Unnamed: 0 DATA INICIAL DATA FINAL        REGIÃO              ESTADO  \
9256        12064   2004-05-09 2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
9257        12065   2004-05-09 2004-05-15  CENTRO OESTE               GOIAS   
9258        12066   2004-05-09 2004-05-15  CENTRO OESTE         MATO GROSSO   
9259        12067   2004-05-09 2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
9260        12068   2004-05-09 2004-05-15      NORDESTE             ALAGOAS   
...           ...          ...        ...           ...                 ...   
78083      120721   2021-04-25 2021-05-01         NORTE             RORAIMA   
78084      120722   2021-04-25 2021-05-01           SUL      SANTA CATARINA   
78085      120723   2021-04-25 2021-05-01       SUDESTE           SAO PAULO   
78086      120724   2021-04-25 2021-05-01      NORDESTE             SERGIPE   
78087      120725   2021-04-25 2021-05-01         NORTE           TOCANTINS   

              PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
9256   GASOLINA COMUM                           128              R$/l   
9257   GASOLINA COMUM                           395              R$/l   
9258   GASOLINA COMUM                           194              R$/l   
9259   GASOLINA COMUM                           166              R$/l   
9260   GASOLINA COMUM                           106              R$/l   
...               ...                           ...               ...   
78083  GASOLINA COMUM                            12              R$/l   
78084  GASOLINA COMUM                           119              R$/l   
78085  GASOLINA COMUM                          1187              R$/l   
78086  GASOLINA COMUM                            34              R$/l   
78087  GASOLINA COMUM                            20              R$/l   

       PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
9256                 2.029                  0.007                 1.990   
9257                 2.025                  0.062                 1.850   
9258                 2.358                  0.066                 2.000   
9259                 2.120                  0.075                 1.970   
9260                 2.090                  0.034                 2.000   
...                    ...                    ...                   ...   
78083                5.364                  0.065                 5.310   
78084                5.051                  0.251                 4.640   
78085                5.265                  0.218                 4.279   
78086                5.586                  0.088                 5.410   
78087                5.751                  0.160                 5.578   

       PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
9256                  2.070                0.318                     0.003   
9257                  2.220                0.296                     0.031   
9258                  2.540                0.472                     0.028   
9259                  2.440                0.325                     0.035   
9260                  2.159                 0.35                     0.016   
...                     ...                  ...                       ...   
78083                 5.530             -99999.0                     0.012   
78084                 5.599             -99999.0                     0.050   
78085                 6.099             -99999.0                     0.041   
78086                 5.699             -99999.0                     0.016   
78087                 6.119             -99999.0                     0.028   

      PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
9256                     1.711                       0.02   
9257                     1.729                      0.036   
9258                     1.886                      0.068   
9259                     1.795                      0.033   
9260                      1.74        

### 8. Qual o preço médio de revenda da gasolina em agosto de 2008?

In [138]:
df_amostra[df_amostra['ANO - MÊS'] == '2008 - 08']['PREÇO MÉDIO REVENDA'].mean()

2.6012444444444447

In [139]:
df_amostra['ANO - MÊS'].info()

<class 'pandas.core.series.Series'>
Int64Index: 23570 entries, 9256 to 78087
Series name: ANO - MÊS
Non-Null Count  Dtype 
--------------  ----- 
23570 non-null  object
dtypes: object(1)
memory usage: 368.3+ KB


In [140]:
df_amostra[df_amostra['ANO - MÊS'] == '2004 - 06']['NÚMERO DE POSTOS PESQUISADOS']

9337     144
9338     387
9339     199
9340     180
9341     108
        ... 
9440    1090
9441    4140
9442     748
9443     817
9444     520
Name: NÚMERO DE POSTOS PESQUISADOS, Length: 108, dtype: int64

In [141]:
df_amostra[df_amostra['ESTADO'] == 'CEARA']

Unnamed: 0 DATA INICIAL DATA FINAL    REGIÃO ESTADO         PRODUTO  \
9262        12070   2004-05-09 2004-05-15  NORDESTE  CEARA  GASOLINA COMUM   
9289        12097   2004-05-16 2004-05-22  NORDESTE  CEARA  GASOLINA COMUM   
9316        12124   2004-05-23 2004-05-29  NORDESTE  CEARA  GASOLINA COMUM   
9343        12151   2004-05-30 2004-06-05  NORDESTE  CEARA  GASOLINA COMUM   
9370        12178   2004-06-06 2004-06-12  NORDESTE  CEARA  GASOLINA COMUM   
...           ...          ...        ...       ...    ...             ...   
77352      119990   2021-03-28 2021-04-03  NORDESTE  CEARA  GASOLINA COMUM   
77530      120168   2021-04-04 2021-04-10  NORDESTE  CEARA  GASOLINA COMUM   
77709      120347   2021-04-11 2021-04-17  NORDESTE  CEARA  GASOLINA COMUM   
77887      120525   2021-04-18 2021-04-24  NORDESTE  CEARA  GASOLINA COMUM   
78066      120704   2021-04-25 2021-05-01  NORDESTE  CEARA  GASOLINA COMUM   

       NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
9262                            321              R$/l                2.148   
9289                            321              R$/l                2.147   
9316                            329              R$/l                2.147   
9343                            317              R$/l                2.148   
9370                            326              R$/l                2.147   
...                             ...               ...                  ...   
77352                           150              R$/l                5.488   
77530                           150              R$/l                5.503   
77709                           152              R$/l                5.503   
77887                           137              R$/l                5.471   
78066                           135              R$/l                5.490   

       DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
9262                   0.080                 1.930                 2.340   
9289                   0.080                 1.930                 2.340   
9316                   0.079                 1.930                 2.340   
9343                   0.080                 1.930                 2.340   
9370                   0.078                 1.930                 2.340   
...                      ...                   ...                   ...   
77352                  0.096                 5.159                 5.899   
77530                  0.124                 5.159                 5.899   
77709                  0.107                 5.190                 5.899   
77887                  0.114                 5.190                 5.859   
78066                  0.103                 5.159                 5.699   

      MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
9262                 0.344                     0.037                    1.804   
9289                 0.338                     0.037                    1.809   
9316                  0.34                     0.037                    1.807   
9343                 0.338                     0.037                     1.81   
9370                  0.33                     0.036                    1.817   
...                    ...                       ...                      ...   
77352             -99999.0                     0.017                 -99999.0   
77530             -99999.0                     0.023                 -99999.0   
77709             -99999.0                     0.019                 -99999.0   
77887             -99999.0                     0.021                 -99999.0   
78066             -99999.0                     0.019                 -99999.0   

      DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
9262                       0.034                     1.715   
9289                       0.035                    1.7281   
9316                       0.035                     1.695   

### 9. Qual o preço médio de revenda da gasolina em maio de 2021 em Ceará?

In [142]:
df['ANO'] = df['DATA FINAL'].apply(lambda ano : '{}'.format(ano.year))
df['MÊS'] = df['DATA FINAL'].apply(lambda mes : '{}'.format(mes.month))
df.head(3)

Unnamed: 0 DATA INICIAL DATA FINAL        REGIÃO            ESTADO  \
0           0   2004-05-09 2004-05-15  CENTRO OESTE  DISTRITO FEDERAL   
1           1   2004-05-09 2004-05-15  CENTRO OESTE             GOIAS   
2           2   2004-05-09 2004-05-15  CENTRO OESTE       MATO GROSSO   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  MARGEM MÉDIA REVENDA  \
0                1.288                  0.016  ...                 0.463   
1                1.162                  0.114  ...                 0.399   
2                1.389                  0.097  ...                 0.419   

   COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                     0.012                    0.825   
1                     0.098                    0.763   
2                     0.070                     0.97   

   DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                        0.11                    0.4201   
1                       0.088                    0.5013   
2                       0.095                    0.5614   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO  ANO - MÊS   ANO MÊS  
0                    0.9666                         0.133  2004 - 05  2004   5  
1                      1.05                         0.115  2004 - 05  2004   5  
2                     1.161                         0.098  2004 - 05  2004   5  

[3 rows x 22 columns]

In [143]:
df[(df['ANO'] == '2021') & (df['MÊS'] == '5')]['PREÇO MÉDIO REVENDA']

78007    5.048
78008    4.625
78009    4.750
78010    4.100
78011    4.074
         ...  
78180    4.582
78181    4.141
78182    4.246
78183    4.464
78184    4.352
Name: PREÇO MÉDIO REVENDA, Length: 178, dtype: float64

In [144]:
df[(df['ANO'] == '2021') & (df['MÊS'] == '5')]['PREÇO MÉDIO REVENDA'].mean()

17.502089887640455

## Obs:
   *Vamos observar o tamanho do arquivo em df.*

In [145]:
type(df['DATA INICIAL'].iloc[0])

pandas._libs.tslibs.timestamps.Timestamp

### 10. Você conseguiria descobrir em qual(quais) estado(s) a gasolina ultrapassou a barreira dos R$ 5,00? E quando isso ocorreu?

In [146]:
df[(df['PRODUTO'] == 'GASOLINA COMUM') & (df['PREÇO MÉDIO REVENDA'] > 5)][['REGIÃO','ESTADO','PREÇO MÉDIO REVENDA','MÊS','ANO']]

REGIÃO          ESTADO  PREÇO MÉDIO REVENDA MÊS   ANO
55563     NORTE            ACRE                5.053   5  2018
55709     NORTE            ACRE                5.035   6  2018
55855     NORTE            ACRE                5.038   6  2018
55873   SUDESTE  RIO DE JANEIRO                5.016   6  2018
56008     NORTE            ACRE                5.031   6  2018
...         ...             ...                  ...  ..   ...
78083     NORTE         RORAIMA                5.364   5  2021
78084       SUL  SANTA CATARINA                5.051   5  2021
78085   SUDESTE       SAO PAULO                5.265   5  2021
78086  NORDESTE         SERGIPE                5.586   5  2021
78087     NORTE       TOCANTINS                5.751   5  2021

[353 rows x 5 columns]

##### Repare que está é uma nova forma de mostrar o data frame. Ainda é o mesmo df, porém, de maneira análoga ao SQL, ele me retornou uma tabela reduzida
##### e somente com os valores especificados por mim.
    Agora farei um teste, tentarei passar esses 353 linhas e 5 colunas para uma novo data frame.
    Observe que, o novo dataFrame está modificado. não se parece com o df que é o original.

In [147]:
teste = pd.DataFrame(df[(df['PRODUTO'] == 'GASOLINA COMUM') & (df['PREÇO MÉDIO REVENDA'] > 5)][['REGIÃO','ESTADO','PREÇO MÉDIO REVENDA','ANO - MÊS']])
teste

REGIÃO          ESTADO  PREÇO MÉDIO REVENDA  ANO - MÊS
55563     NORTE            ACRE                5.053  2018 - 05
55709     NORTE            ACRE                5.035  2018 - 06
55855     NORTE            ACRE                5.038  2018 - 06
55873   SUDESTE  RIO DE JANEIRO                5.016  2018 - 06
56008     NORTE            ACRE                5.031  2018 - 06
...         ...             ...                  ...        ...
78083     NORTE         RORAIMA                5.364  2021 - 05
78084       SUL  SANTA CATARINA                5.051  2021 - 05
78085   SUDESTE       SAO PAULO                5.265  2021 - 05
78086  NORDESTE         SERGIPE                5.586  2021 - 05
78087     NORTE       TOCANTINS                5.751  2021 - 05

[353 rows x 4 columns]

### 11. Qual a média de preço dos estados da região sul em 2012?

In [150]:
df["REGIÃO"].unique()

array(['CENTRO OESTE', 'NORDESTE', 'NORTE', 'SUDESTE', 'SUL'],
      dtype=object)

In [152]:
teste_aux = df[df['REGIÃO'] == 'SUL']
teste_aux

Unnamed: 0 DATA INICIAL DATA FINAL REGIÃO             ESTADO  \
24             24   2004-05-09 2004-05-15    SUL             PARANA   
25             25   2004-05-09 2004-05-15    SUL  RIO GRANDE DO SUL   
26             26   2004-05-09 2004-05-15    SUL     SANTA CATARINA   
51             51   2004-05-16 2004-05-22    SUL             PARANA   
52             52   2004-05-16 2004-05-22    SUL  RIO GRANDE DO SUL   
...           ...          ...        ...    ...                ...   
78151      120789   2021-04-25 2021-05-01    SUL  RIO GRANDE DO SUL   
78154      120792   2021-04-25 2021-05-01    SUL     SANTA CATARINA   
78173      120811   2021-04-25 2021-05-01    SUL             PARANA   
78178      120816   2021-04-25 2021-05-01    SUL  RIO GRANDE DO SUL   
78181      120819   2021-04-25 2021-05-01    SUL     SANTA CATARINA   

                PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
24     ETANOL HIDRATADO                           760              R$/l   
25     ETANOL HIDRATADO                           689              R$/l   
26     ETANOL HIDRATADO                           506              R$/l   
51     ETANOL HIDRATADO                           769              R$/l   
52     ETANOL HIDRATADO                           674              R$/l   
...                 ...                           ...               ...   
78151       OLEO DIESEL                           111              R$/l   
78154       OLEO DIESEL                            61              R$/l   
78173   OLEO DIESEL S10                           142              R$/l   
78178   OLEO DIESEL S10                           148              R$/l   
78181   OLEO DIESEL S10                           100              R$/l   

       PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  MARGEM MÉDIA REVENDA  \
24                   0.922                  0.083  ...                 0.255   
25                   1.224                  0.143  ...                 0.332   
26                   1.229                  0.077  ...                 0.416   
51                   0.922                  0.088  ...                 0.254   
52                   1.220                  0.142  ...                 0.324   
...                    ...                    ...  ...                   ...   
78151                4.079                  0.191  ...              -99999.0   
78154                4.064                  0.206  ...              -99999.0   
78173                4.044                  0.166  ...              -99999.0   
78178                4.140                  0.236  ...              -99999.0   
78181                4.141                  0.187  ...              -99999.0   

       COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
24                        0.090                    0.667   
25                        0.117                    0.892   
26                        0.063                    0.813   
51                        0.095                    0.668   
52                        0.116                    0.896   
...                         ...                      ...   
78151                     0.047                 -99999.0   
78154                     0.051                 -99999.0   
78173                     0.041                 -99999.0   
78178                     0.057                 -99999.0   
78181                     0.045                 -99999.0   

       DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
24                          0.105                    0.4357   
25                           0.09                     0.644   
26                          0.097                      0.57   
51                            0.1                    0.4337   
52                          0.088                     0.644   
...                           ...                       ...   
78151                    -99999.0                  -99999.0   
78154                    -99999.0                  -99999.0   
7817

### Agora como faço para saber o nome dos estados da região SUL?
    Simples, com a função unique(), ela retorna somente os únicos valores dos nomes dos estados da região sul.

In [153]:
teste_aux['ESTADO'].unique()

array(['PARANA', 'RIO GRANDE DO SUL', 'SANTA CATARINA'], dtype=object)

In [154]:
teste_aux['ANO'].unique()

array(['2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011',
       '2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019',
       '2020', '2021'], dtype=object)

### Agora respondendo a pergunta 11 !

In [158]:
teste_aux[teste_aux['ANO'] == '2012']['PREÇO MÉDIO REVENDA']

1428     2.046
1429     2.414
1430     2.423
1455     2.034
1456     2.408
         ...  
12948    2.190
12949    2.168
12974    2.091
12975    2.186
12976    2.170
Name: PREÇO MÉDIO REVENDA, Length: 780, dtype: float64

### Agora o preço médio de todos os valores do ano de 2012 para a região do sul.

In [156]:
teste_aux2 = df[(df["DATA FINAL"].apply(lambda x: x.year) == 2012)]
teste_aux2[teste_aux2["REGIÃO"] == "SUL"]["PREÇO MÉDIO REVENDA"].mean()

9.842052564102566

### 12. Você conseguiria obter uma tabela contendo a variação percentual ano a ano para o estado do Ceará?

In [161]:
df['ESTADO'].unique()

array(['DISTRITO FEDERAL', 'GOIAS', 'MATO GROSSO', 'MATO GROSSO DO SUL',
       'ALAGOAS', 'BAHIA', 'CEARA', 'MARANHAO', 'PARAIBA', 'PERNAMBUCO',
       'PIAUI', 'RIO GRANDE DO NORTE', 'SERGIPE', 'ACRE', 'AMAPA',
       'AMAZONAS', 'PARA', 'RONDONIA', 'RORAIMA', 'TOCANTINS',
       'ESPIRITO SANTO', 'MINAS GERAIS', 'RIO DE JANEIRO', 'SAO PAULO',
       'PARANA', 'RIO GRANDE DO SUL', 'SANTA CATARINA'], dtype=object)

In [162]:
df.head(2)

Unnamed: 0 DATA INICIAL DATA FINAL        REGIÃO            ESTADO  \
0           0   2004-05-09 2004-05-15  CENTRO OESTE  DISTRITO FEDERAL   
1           1   2004-05-09 2004-05-15  CENTRO OESTE             GOIAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  ...  MARGEM MÉDIA REVENDA  \
0                1.288                  0.016  ...                 0.463   
1                1.162                  0.114  ...                 0.399   

   COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                     0.012                    0.825   
1                     0.098                    0.763   

   DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                        0.11                    0.4201   
1                       0.088                    0.5013   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO  ANO - MÊS   ANO MÊS  
0                    0.9666                         0.133  2004 - 05  2004   5  
1                      1.05                         0.115  2004 - 05  2004   5  

[2 rows x 22 columns]